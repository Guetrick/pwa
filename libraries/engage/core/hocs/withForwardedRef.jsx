import React, { forwardRef } from 'react';

/**
 * Provides a wrapper for components that utilize forwarded refs. It accepts a ref via props,
 * and passes it down as a prop called forwardedRef.
 * @param {Function} WrappedComponent The react component to wrap.
 * @returns {JSX}
 */
export function withForwardedRef(WrappedComponent) {
  /**
   * The actual HOC.
   * @param {Object} props The component props.
   * @param {Object} ref The forwarded ref.
   * @returns {JSX}
   */
  const handle = (props, ref) => <WrappedComponent {...props} forwardedRef={ref} />;

  const name = WrappedComponent.displayName || WrappedComponent.name;
  handle.displayName = `withForwardedRef(${name})`;

  return forwardRef(handle);
}
