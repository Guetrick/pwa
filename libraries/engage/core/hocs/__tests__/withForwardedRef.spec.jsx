import React, { createRef } from 'react';
import { renderIntoDocument } from 'react-dom/test-utils';
import { shallow } from 'enzyme';
import { withForwardedRef } from '../withForwardedRef';

/* eslint-disable require-jsdoc, react/prop-types */
describe('withForwardedRef', () => {
  it('should render', () => {
    const Comp = ({ children, className, forwardedRef }) => (
      <div className={className} ref={forwardedRef}>
        {children}
      </div>
    );
    const WrappedComp = withForwardedRef(Comp);
    const ref = createRef();
    const wrapper = shallow((
      <WrappedComp className="foo" ref={ref} >
        Testing 123
      </WrappedComp>
    ));
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.name()).toBe('Comp');
  });

  it('should render with a custom displayName', () => {
    const Comp = ({ children, className, forwardedRef }) => (
      <div className={className} ref={forwardedRef}>
        {children}
      </div>
    );
    Comp.displayName = 'FooBar';
    const WrappedComp = withForwardedRef(Comp);
    const ref = createRef();
    const wrapper = shallow((
      <WrappedComp className="foo" ref={ref} >
        Testing 123
      </WrappedComp>
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.name()).toBe(Comp.displayName);
  });

  it('should create a ref that points to a HTMLDivElement', () => {
    const Comp = ({ children, className, forwardedRef }) => (
      <div className={className} ref={forwardedRef}>
        {children}
      </div>
    );
    const WrappedComp = withForwardedRef(Comp);
    const ref = createRef();
    renderIntoDocument((
      <WrappedComp className="foo" ref={ref} >
        Testing 123
      </WrappedComp>
    ));
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
  });
});

/* eslint-enable require-jsdoc, react/prop-types */
