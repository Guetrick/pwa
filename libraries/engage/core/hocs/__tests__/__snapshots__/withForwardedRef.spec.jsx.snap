// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage > core > hocs > withForwardedRef should inject the ref with a custom prop name 1`] = `
<MockComponent
  className="foo"
  custom={
    Object {
      "current": null,
    }
  }
>
  Testing 123
</MockComponent>
`;

exports[`engage > core > hocs > withForwardedRef should render 1`] = `
<MockComponent
  className="foo"
  forwardedRef={
    Object {
      "current": null,
    }
  }
>
  Testing 123
</MockComponent>
`;
